kernel/packages/voice-chat-codec/worker.ts

```diff
@@ -0,0 +1,157 @@
+import { VoiceChatWorkerResponse, RequestTopic, ResponseTopic } from './types'
+import { Resampler } from './resampler'
+declare var self: WorkerGlobalScope & any
+
+declare function postMessage(message: any, transferables: any[]): void
+
+declare const libopus: any
+
+self.LIBOPUS_WASM_URL = 'libopus.wasm'
+
+importScripts('libopus.wasm.js')
+
+type CodecWorklet = {
+  working: boolean
+  lastWorkTime: number
+  destroy: () => any
+}
+
+type EncoderWorklet = {
+  encoder: any
+} & CodecWorklet
+
+type DecoderWorklet = {
+  decoder: any
+} & CodecWorklet
+
+function getSampleRate(e: MessageEvent) {
+  return e.data.sampleRate ? e.data.sampleRate : 24000
+}
+
+const encoderWorklets: Record<string, EncoderWorklet> = {}
+const decoderWorklets: Record<string, DecoderWorklet> = {}
+
+function startWorklet<T extends CodecWorklet, O extends Uint8Array | Float32Array>(
+  streamId: string,
+  worklet: T,
+  outputFunction: (worklet: T) => O,
+  messageBuilder: (output: O, streamId: string) => VoiceChatWorkerResponse
+) {
+  worklet.working = true
+
+  function doWork() {
+    worklet.lastWorkTime = Date.now()
+
+    let output = outputFunction(worklet)
+
+    if (output) {
+      if (output instanceof Uint8Array) {
+        output = Uint8Array.from(output) as O
+      } else {
+        output = Float32Array.from(output) as O
+      }
+
+      postMessage(messageBuilder(output, streamId), [output.buffer])
+      setTimeout(doWork, 0)
+    } else {
+      worklet.working = false
+    }
+  }
+
+  setTimeout(doWork, 0)
+}
+
+//Encoder(channels, samplerate, bitrate, frame_size, voice_optimization)
```

Comment:
Maybe we can remove this comment or at least explain what is this supposed to help with.